### ISSUE TITLE
Provide speedups for the Robot Framework parser using cython

### ISSUE THREAD

**Issue by fabioz** [robotframework/robotframework#4565]
The speed of the parser can be a deal breaker when dealing even with medium sized files, so, it'd be really nice to have speed-up modules to make parsing faster.

**Comment by pekkaklarck** [comment]
Sounds like a good idea. Do you have any estimate how big task this would be? This most likely will require external funding or contributions.

**Comment by fabioz** [comment]
It really depends on how deep down to go... at the basic level just having cython with python type annotations should be reasonably straightforward to do now that Python 3 is a requisite and should be a good first step.

The steps here are:
- adding the cython dependency
- adding some more commands in setup.py to have cython compile the needed modules
- add python type annotations to the modules to accelerate
- create some scaffolding to load those modules when they're available instead of the pure-python modules
- create binaries for each different platform (the main issue with cython is that the binary it produces is compatible only for the given python version, so, we need one binary for each new platform -- i.e.: it's a matrix of OS+arch (for instance win amd64 or mac aarch64) and the python version (3.6, 3.7, 3.8, ...).

After having that in place it may be possible to further optimizations for cython... What I usually do is create some code which generates the cython code to be compiled using some comments (see: https://github.com/fabioz/PyDev.Debugger/blob/pydev_debugger_2_9_3/_pydevd_bundle/pydevd_frame.py#L143 for a reference of what I mean), but this can definitely be a 2nd step to squeeze more performance out of it (doing the first part would already provide a nice speed bump... from my usage of cython, 20-30 times faster with the initial approach would be straightforward and the 2nd step could raise it up to 50-80 time faster).

There are other alternatives though:
- having code in a different language and producing a single binary for each OS+arch and then using ctypes to deal with it (the downside is having to keep separate code in a different language).
- Using mypyc (the downside is that it seems less stable and with cython there's more room for other optimizations down the line, such as releasing the GIL so that it's possible to take advantage of doing parses in parallel).
- Using numba (although for the use case I think numba is not an ideal match as it's better at single algorithms and for this use case we have more Python structure).

Note that it'd probably be nice if https://github.com/robotframework/robotframework/issues/4564 would be tackled before that.

Time-wise I'd say that the first part could be done in around 2 weeks (working on it full time) to have an initial working version and maybe one more week for polish and doing profiles to squeeze more performance out of it.

I can probably help on doing more improvements on the cython side once the basic integration is done... regarding changes to integrate cython itself to the build, I'm not sure even how the build is done right now, so, it may be harder to contribute (I went over to https://github.com/robotframework/robotframework/tree/master/.github/workflows and didn't really find the release step there -- is this a manual process you do?).

**Comment by pekkaklarck** [comment]
Yeah, having the parser available as a separate package (#4564) is definitely a precondition for this. This is also so big investment, and doesn't provide benefits for those who are fine with the current parser speed, that this will require a separate decision by the foundation board.

Our build process is defined in BUILD.rst. The overall process is manual, but individual steps are automated using `invoke`.

Is the Cython code still compatible with normal Python so that we could use a single code base?

**Comment by fabioz** [comment]
> Is the Cython code still compatible with normal Python so that we could use a single code base?

In general yes (some more advanced changes may require code which ends up not being plain python, but then what I usually do is add some instructions to the original file as comments and generate the file to be fed to cython as a separate build -- this would be step 2 though, for step 1 only plain python code is expected).

> Our build process is defined in BUILD.rst. The overall process is manual, but individual steps are automated using `invoke`.

I'd say that doing automated builds may be the first requirement to doing the cython builds then (my usual approach is that when I create a tag with a specified format it'll create a build. Reference: https://github.com/robocorp/robotframework-lsp/blob/master/.github/workflows/release-robotframework-lsp.yml). As you'll have to do a build matrix of OS+arch/python version it just doesn't scale doing it manually...

**Comment by bollwyvl** [comment]
To expound on the `mypy` tack: 

It's rather difficult to anticipate the performance difference without a proper bake-off (twice the dev work), but from an end-user standpoint, the underlying code is still plain-old-python, and will run, unmodified, on any python runtime, which is a _distinct_ advantage over cython.

In any event, i feel performance gains can only be realized and maintained with consistent benchmarking, which would suggest tools like: 
- [asv](https://asv.readthedocs.io) to demonstrate the issue over time
  - ideally against _historical_ versions against different pythons
- [pyinstrument](https://github.com/joerick/pyinstrument) for generating good instrumentation data
- [speedscope](https://github.com/jlfwong/speedscope) for visualizing
  - of note... speedscope JSON would be an _amazing_ transformation target for robot XML 

As with cython, `mypyc` _would_ require building and pushing out many more wheels to deliver the performance benefits to folk, but the [recommended workflow](https://mypyc.readthedocs.io/en/latest/getting_started.html#recommended-workflow) suggests pushing out a "fallback" `py3-none-any.whl`, which would suffer no additional performance penalty (modulo any complexity of "fancy" types), but _crucially_ no added install or developer complexity.

Total aside: the funny thing is I remember discussing pretty much _exactly_ these topics, even before `mypyc` existed, with @datakurre and @rasjani a number of RobosCon ago while trudging through the lovely January Helsinki weather. Sorry I can't be joining you folks this time around!

**Comment by jiridanek** [comment]
Somewhat related to
* https://github.com/robotframework/robotframework/issues/2592

Does PyPy help any? Or what about the experimental [CPython 3.13 JIT](https://docs.python.org/3/whatsnew/3.13.html#whatsnew313-jit-compiler)? In my view it would be a good start to try to achieve (and quantify) some such easy wins first.
